fast_llm/engine/training/config.py

```diff
@@ -90,71 +141,109 @@ class WandbConfig(Config):
         hint=FieldHint.feature,
         valid=skip_valid_if_none(check_field(Assert.gt, 0)),
     )
 
+    def get_save_config(self, path: pathlib.Path):
+        raise NotImplementedError()
 
-def _validate_script(value):
-    if isinstance(value, str):
-        value = shlex.split(value)
-    Assert.geq(len(value), 1)
-    return value
+    def to_delete(self, iterations: list[int]):
+        if not self.keep:
+            return []
+        # Ignore checkpoints that aren't supposed to be there.
+        iterations = [iteration for iteration in iterations if self.enabled(iteration)]
+        # Ignore excluded checkpoints.
+        if self.keep_every:
+            iterations = [iteration for iteration in iterations if self.get_count(iteration) % self.keep_every != 0]
+        # Exclude the last `keep`.
+        return iterations[: -self.keep]
 
 
 @config_class()
-class CallbackConfig(Config):
-    script: list[str] | None = Field(
-        default=None,
-        desc="Shell script to run after.",
-        hint=FieldHint.feature,
-        valid=skip_valid_if_none(_validate_script),
-    )
-    environment: dict[str, str] = Field(
-        default_factory=dict,
-        desc="Environment variables to add to the script.",
-        hint=FieldHint.feature,
+class CheckpointConfig(CheckpointBaseConfig):
+    _abstract = False
+    save_name: typing.ClassVar[str] = "checkpoint"
+    # TODO v0.2: Rename to `checkpoint` so we don't need this extra variable?
+    directory_name = "checkpoints"
+    interval = FieldUpdate(
+        desc="The number of training iterations between each checkpoint." " Setting to None will disable checkpoints."
     )
-
-    def run(self):
-        if self.script is not None:
-            environment = os.environ.copy()
-            environment.update(self.environment)
-            subprocess.Popen(self.script, env=environment)
+    offset = FieldUpdate(desc="Offset for the first checkpoint.")
+    callback: CallbackConfig = FieldUpdate(desc="Callback (shell script) to run after checkpoint.")
+    keep: int | None = FieldUpdate(default=5)
+
+    def get_save_config(self, path: pathlib.Path):
+        return CheckpointSaveConfig(
+            path=path,
+            format=CheckpointFormat.distributed,
+            model_weights=True,
+            optimizer_state=True,
+        )
 
 
 @config_class()
-class ExportConfig(get_interval_config_class("export")):
-    callback: CallbackConfig = Field(
-        default_factory=CallbackConfig,
-        desc="Callback (shell script) to run after export.",
-        hint=FieldHint.core,
+class ExportConfig(CheckpointBaseConfig, CheckpointConfigBase, CheckpointStateConfigBase, CheckpointSaveConfigBase):
+    _abstract = False
+    save_name: typing.ClassVar[str] = "export"
+    directory_name = "export"
+    interval = FieldUpdate(
+        desc="The number of training iterations between each export." " Setting to None will disable exports."
```

Comment:
Can this interval be incompatible with the checkpointing interval (i.e., not an exact multiple or divisor of the checkpointing interval)?

Reply:
Yes it doesn't matter anymore because checkpoints and exports are completely independent.